### PR TITLE
CDAP-18890 stop programs in an async fashion through TMS

### PR DESCRIPTION
**Pull request by albertshau**
Refactored program stop so that a stopping message is written out
when a program stop is issued instead of trying to actually stop
the run. After the stopping state is recorded, a new subscriber
will perform the actual program stop.

As part of this, added a kill method to the ProgramController, to
differentiate between a graceful stop and a force kill, and to
better match the TwillController API that already exists.
Changed the RemoteTwillController so that the graceful stop simply
waits for the RuntimeClient to stop the program itself.
Also changed the KubeTwillController so that it specifies a
graceful stop argument accordingly when deleting
jobs/deployments/statefulsets

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java [hunk 1]
```diff
@@ -766,111 +734,35 @@ public void stop(ProgramId programId, @Nullable String runId) throws Exception {
...
-
-        if (!pendingStops.isEmpty()) {
-          TimeUnit.MILLISECONDS.sleep(200);
-        }
+    for (Map.Entry<ProgramRunId, RunRecordDetail> activeRunRecord : activeRunRecords.entrySet()) {
+      ProgramRunId activeRunId = activeRunRecord.getKey();
+      RunRecordDetail runRecord = activeRunRecord.getValue();
+
+      // Check if the program is actually started from workflow and the workflow is running
+      if (runRecord != null && runRecord.getProperties().containsKey("workflowrunid")
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java]: nit: workflowrunid can be a constant

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java [hunk 2]
```diff
@@ -766,111 +734,35 @@ public void stop(ProgramId programId, @Nullable String runId) throws Exception {
...
+        && runRecord.getStatus().equals(ProgramRunStatus.RUNNING)) {
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java:756]: If stop is issued on a program that is running then we are throwing a bad request exception? 🤔

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramStopSubscriberService.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Subscribes to the Constants.AppFabric.PROGRAM_STATUS_RECORD_EVENT_TOPIC topic, which gets published to
+ * after the state store has been updated. Looks for stopping messages and performs the actual program stop.
+ */
+public class ProgramStopSubscriberService extends AbstractNotificationSubscriberService {
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramStopSubscriberService.java:52]: Just so I am understanding this correctly, this service is required to handle stops for non-dataproc envs?
For dataproc, we already have changes in DefaultRuntimeJob on the client side and the periodic service in app-fabric to scan for programs running beyond terminate time. How would this change affect dataproc case?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java (hunk 2 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java:756]: this is existing code, if somebody tries to stop a program that is part of a workflow run they get a 400 response. 

For example, if they try to stop the 'phase-1' Spark program that is part of the 'DataPipelineWorkflow', they get a 400 asking them to stop the Workflow instead.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramStopSubscriberService.java (hunk 3 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramStopSubscriberService.java:52]: Right, non-remote jobs still need to be stopped. The relevant change is in the RemoteExecutionTwillController.java, where a stop() call just waits for the DefaultRuntimeJob to do its thing and shut itself down. This is instead of calling remoteProcessController.terminate(), which would kill the Dataproc job.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 4]
```diff
@@ -298,13 +296,45 @@ public void performRunLevelStop(HttpRequest request, HttpResponder responder,
...
-    lifecycleService.stop(program, runId);
+    int gracefulShutdownSecsInt;
+    try {
+      gracefulShutdownSecsInt = validateAndGetGracefulShutdownSecsInt(gracefulShutdownSecs);
+    } catch (IllegalArgumentException e) {
+      responder.sendJson(HttpResponseStatus.BAD_REQUEST, e.getMessage());
+      return;
+    }
+    try {
+      RunIds.fromString(runId);
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:311]: Is there a way to also validate program type and program id?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java [hunk 5]
```diff
@@ -766,111 +734,35 @@ public void stop(ProgramId programId, @Nullable String runId) throws Exception {
...
+                                            int gracefulShutdownSecs) throws Exception {
     accessEnforcer.enforce(programId, authenticationContext.getPrincipal(), ApplicationPermission.EXECUTE);
 
-    // See if the program is running as per the runtime service
-    Map<RunId, RuntimeInfo> runtimeInfos = findRuntimeInfo(programId, runId);
     Map<ProgramRunId, RunRecordDetail> activeRunRecords = getActiveRuns(programId, runId);
 
-    if (runtimeInfos.isEmpty() && activeRunRecords.isEmpty()) {
+    if (activeRunRecords.isEmpty()) {
       // Error out if no run information from runtime service and from run record
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java]: Should this be changed to `// Error out if no run information from run record`?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java [hunk 6]
```diff
@@ -766,111 +734,35 @@ public void stop(ProgramId programId, @Nullable String runId) throws Exception {
...
-        }
+    for (Map.Entry<ProgramRunId, RunRecordDetail> activeRunRecord : activeRunRecords.entrySet()) {
+      ProgramRunId activeRunId = activeRunRecord.getKey();
+      RunRecordDetail runRecord = activeRunRecord.getValue();
+
+      // Check if the program is actually started from workflow and the workflow is running
+      if (runRecord != null && runRecord.getProperties().containsKey("workflowrunid")
+        && runRecord.getStatus().equals(ProgramRunStatus.RUNNING)) {
+        String workflowRunId = runRecord.getProperties().get("workflowrunid");
+        throw new BadRequestException(String.format("Cannot stop the program '%s' started by the Workflow " +
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java]: `Cannot stop the **run** '%s'`?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SystemProgramManagementService.java [hunk 7]
```diff
@@ -133,7 +133,7 @@ private void startPrograms(Map<ProgramId, Arguments> enabledProgramsMap) {
   private void stopProgram(ProgramRunId programRunId) {
     LOG.debug("Stopping program run {} ", programRunId);
     try {
-      programLifecycleService.stop(programRunId.getParent(), programRunId.getRun());
+      programLifecycleService.stop(programRunId.getParent(), programRunId.getRun(), 30);
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SystemProgramManagementService.java]: Why specifically 30 secs?

File: cdap-common/src/main/java/io/cdap/cdap/gateway/handlers/ConfigHandler.java [hunk 8]
```diff
@@ -53,6 +54,7 @@ public void configCDAP(@SuppressWarnings("UnusedParameters") HttpRequest request
                          @DefaultValue("json") @QueryParam("format") String format) throws IOException {
     if ("json".equals(format)) {
       responder.sendJson(HttpResponseStatus.OK, GSON.toJson(configService.getCConf()));
+      TimeUnit.MINUTES.toMillis(5);
```
Comment (snehalakshmisha) [inline cdap-common/src/main/java/io/cdap/cdap/gateway/handlers/ConfigHandler.java]: Why is this required?

**Review by snehalakshmisha — COMMENTED** [review]
Minor comments, otherwise LGTM

**Review by snehalakshmisha — COMMENTED** [review]
We probably also need to enhance ProgramLifecycleHttpHandlerTest and ProgramLifecycleServiceTest

File: cdap-common/src/main/java/io/cdap/cdap/gateway/handlers/ConfigHandler.java (hunk 8 above)
Comment (albertshau) [reply cdap-common/src/main/java/io/cdap/cdap/gateway/handlers/ConfigHandler.java]: oops :) will remove

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 4 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:311]: it's validated when creating ProgramType and ProgramId objects

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SystemProgramManagementService.java (hunk 7 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SystemProgramManagementService.java]: hm this should be integer max actually, will change

File: cdap-common/src/main/java/io/cdap/cdap/gateway/handlers/ConfigHandler.java [hunk 9]
```diff
@@ -28,6 +28,7 @@
 import io.netty.handler.codec.http.HttpResponseStatus;
 
 import java.io.IOException;
+import java.util.concurrent.TimeUnit;
```
Comment (snehalakshmisha) [inline cdap-common/src/main/java/io/cdap/cdap/gateway/handlers/ConfigHandler.java]: Seems like you missed removing this

**Review by snehalakshmisha — APPROVED** [review]
